building/glass/upload.py

```diff
@@ -123,20 +121,40 @@ def gs_rsync(local_path: str, remote_path: str, boto_path: str):
 """.lstrip()
 
 
-def perform_uploads(uploads: dict, branch: str) -> None:
+def upload_gs(staging, root: str, branch_config: aptbranch.Config):
+    branch = branch_config.name
+    if not GS_BRANCH_REGEX.match(branch):
+        raise Exception("not an uploadable branch: %s" % branch)
     if not os.path.exists(BOTO_PATH):
         raise Exception("you need to put the GCP service account private key file into /homeworld/boto-key")
+    botoconfig_name = os.path.join(staging, "boto.config")
+    with open(botoconfig_name, "w") as bout:
+        with open(BOTO_PATH, "r") as f:
+            project_id = json.load(f)["project_id"]
+        bout.write(BOTO_TEMPLATE % (BOTO_PATH, project_id))
+        bout.flush()
+    gs_rsync(root, branch, botoconfig_name)
+
+
+def upload_rsync(staging, root: str, branch_config: aptbranch.Config):
+    target_dir = os.path.join(branch_config.upload_config["dir"], branch_config.name)
+    target = "%s@%s:%s" % (branch_config.upload_config["username"], branch_config.upload_config["host"], target_dir)
```

Comment:
It might be worth erroring out here if any special characters are included in these fields that could be problematic for rsync's parsing of its parameter? For example, a `-` at the start of the username.

Reply:
I've added "--", which should take care of the dash problem. I also combined the username and host in the config and checked if it contains a colon. Is there anything I missed?

Reply:
I don't know. I usually try to use a whitelist instead of a blacklist, because trying to blacklist format violations is generally a losing game.

However, since we can trust that anyone who can modify the branches.yaml is probably not evil anyway, this might not matter.